File: sdk/src/main/java/com/libreshockwave/cast/FilmLoopInfo.java
package com.libreshockwave.cast;

import com.libreshockwave.io.BinaryReader;

import java.nio.ByteOrder;

/**
 * Film loop-specific cast member data.
 */
public record FilmLoopInfo(
    int regX,
    int regY,
    int width,
    int height,
    boolean center,
    boolean crop,
    boolean sound,
    boolean loops
) implements Dimensioned {

    public static FilmLoopInfo parse(byte[] data) {
        if (data == null || data.length < 11) {
            return new FilmLoopInfo(0, 0, 0, 0, false, true, false, true);
        }

        BinaryReader reader = new BinaryReader(data, ByteOrder.BIG_ENDIAN);

        int regY = reader.readU16();
        int regX = reader.readU16();
        int height = reader.readU16();
        int width = reader.readU16();
        reader.skip(3); // unknown (3 bytes)
        int flags = reader.readU8();

        boolean center = (flags & 0b1) != 0;
        boolean crop = (flags & 0b10) == 0;
        boolean sound = (flags & 0b1000) != 0;
        boolean loops = (flags & 0b100000) == 0;

        return new FilmLoopInfo(regX, regY, width, height, center, crop, sound, loops);
    }
}

File: cpp/src/cast/FilmLoopInfo.cpp
#include "libreshockwave/cast/FilmLoopInfo.hpp"

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::cast {

int FilmLoopInfo::width() const { return rectRight - rectLeft; }
int FilmLoopInfo::height() const { return rectBottom - rectTop; }
int FilmLoopInfo::regX() const { return -rectLeft; }
int FilmLoopInfo::regY() const { return -rectTop; }

FilmLoopInfo FilmLoopInfo::parse(const std::vector<std::uint8_t>& data) {
    if (data.size() < 11) {
        return FilmLoopInfo{0, 0, 0, 0, false, true, false, true};
    }

    io::BinaryReader reader(data, io::ByteOrder::BigEndian);
    const int rectTop = reader.readI16();
    const int rectLeft = reader.readI16();
    const int rectBottom = reader.readI16();
    const int rectRight = reader.readI16();
    reader.skip(3);
    const int flags = reader.readU8();

    return FilmLoopInfo{
        rectTop,
        rectLeft,
        rectBottom,
        rectRight,
        (flags & 0b1) != 0,
        (flags & 0b10) == 0,
        (flags & 0b1000) != 0,
        (flags & 0b100000) == 0
    };
}

} // namespace libreshockwave::cast
